web/src/main/proto/spine/web/subscriptions.proto

```diff
@@ -0,0 +1,181 @@
+option java_multiple_files = true;
+
+import "google/protobuf/any.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+import "spine/base/error.proto";
+import "spine/client/subscription.proto";
+import "spine/core/ack.proto";
+
+// A subscription initiated by a web client.
+//
+// The subscription may have `extra` properties required by the web client.
+//
+message WebSubscription {
+
+    client.Subscription subscription = 1;
+
+    repeated google.protobuf.Any extra = 2;
+}
+
+// A request to subscribe to the given topics.
+message Subscribe {
+
+    // Topics to which to subscribe.
+    repeated client.Topic topic = 1;
+
+    // The requested minimum lifetime of the subscriptions.
+    //
+    // As the clients cannot be trusted to reliably cancel all the subscriptions, the server puts
+    // a limit on the subscription lifetime. When a subscription exceeds the expected lifetime,
+    // the server *may* cancel it.
+    //
+    // In order to prolong subscription lifetime, the client must send the `KeepUp` requests.
+    //
+    // This value is a requested value. The server may set another lifetime to the subscriptions.
+    // The actual "valid through" time is returned by the server in response to this request.
+    //
+    google.protobuf.Duration lifespan = 2;
+}
+
+// The response to the `Subscribe` request.
+message Subscribing {
+
+    // The subscription results list.
+    //
+    // For each subscription topic, this is either a `Subscription`, or an error which prohibits
+    // creating such a subscription.
+    //
+    // The order of the results corresponds to the order of topics in the `Subscribe` request.
+    //
+    repeated SubscriptionOrError result = 1;
+
+    // For all the subscriptions in the response, this is the time after which the subscription
+    // will expire.
+    //
+    // An expired subscription may be cancelled by the server.
+    //
+    google.protobuf.Timestamp valid_thru = 2;
+}
+
+// Either a `spine.client.Subscription` or an error that occurred when creating the subscription.
+message SubscriptionOrError {
+
+    oneof payload {
+        WebSubscription subscription = 1;
+        base.Error error = 2;
+    }
+}
+
+// A request to prolong the life of the given subscriptions.
+message KeepUp {
+
+    // IDs of the subscriptions to keep up.
+    repeated client.SubscriptionId subscription = 1;
+
+    // How much time does the client require.
+    //
+    // The client requests that the server add this amount of time to the subscription life.
+    // The server may add this time or another duration. The updated subscription expiration date
+    // is provided in the response.
+    //
+    // If the client wishes to *reduce* the lifetime of the subscription, it can do so by providing
+    // a negative duration.
+    //
+    google.protobuf.Duration prolong_by = 2;
+}
+
+// The response to the `KeepUp` request.
+message KeepingUp {
```

Comment:
What if we name all responses and outcomes as events? I mean, the requests are already very commanding.

E.g. —

`SubscriptionsKeptUp`,
`SingleSubscriptionKeptUp`,
`SubscriptionsCancelled`.

That would eliminate all the dance with "KeepingUp" and "Cancelling" — which are really the same `Outcome`s, but "..damn it, the word is used already". And honestly, to me "Cancelling" is a process, not a result, or a bunch of results.

Reply:
I've applied this suggestion partially.
The "top-level" types, i.e. `Subscribing`, `KeepingUp`, and `Cancelling` are renamed to sound more event-ish. However, I don't want to use `SingleSubscriptionKeptUp` because it's a bit misleading. It might be kept up, or it not be due to an error.

In a normal command-event schema, we would throw a rejection here. I guess, the HTTP analog of that is an error code. However, the point here is to compose operations on multiple subscriptions under a single request/response pair. Thus, we return an "OK" response with an error as one of the outcomes.